File: src/game_solver/agents/tic_tac_toe/mcts.py
from __future__ import annotations

import math
import random

from game_solver.games.tic_tac_toe import (
    TicTacToeAction,
    TicTacToeAgent,
    TicTacToeState,
)

from .engine import TicTacToeEngine, TicTacToeEngineAction, TicTacToeEngineState


class MCTSNode:
    def __init__(
        self,
        state: TicTacToeEngineState,
        parent: MCTSNode | None = None,
        action: TicTacToeEngineAction | None = None,
    ):
        self.state = state
        self.parent = parent
        self.action = action

        self.children: list[MCTSNode] = []
        self.untried_actions: list[TicTacToeEngineAction] = []

        # Sum of outcomes from state.current_player's perspective.
        self.value_sum = 0.0
        self.visit_count = 0


class TicTacToeMCTSAgent(TicTacToeAgent):
    """UCT を用いた Monte Carlo Tree Search Agent。"""

    def __init__(
        self,
        simulations: int = 1000,
        exploration_weight: float = math.sqrt(2.0),
        seed: int | None = None,
    ) -> None:
        if simulations <= 0:
            raise ValueError("simulations must be positive")
        if exploration_weight < 0:
            raise ValueError("exploration_weight must be non-negative")

        self.simulations = simulations
        self.exploration_weight = exploration_weight

        self._engine = TicTacToeEngine()
        self._rg = random.Random(seed)

    def select_action(self, state: TicTacToeState) -> TicTacToeAction:
        engine_state = self._engine.from_state(state)
        actions = self._engine.legal_actions(engine_state)

        if not actions:
            raise ValueError("No legal actions available.")

        root = self._make_node(engine_state)

        for _ in range(self.simulations):
            node = root

            # 1. Selection: 探索価値の高い葉ノードまで降りていく
            node = self._select(node)

            # 2. Expansion: 未探索の子ノードがあれば1つ展開する
            if node.untried_actions:
                node = self._expand(node)

            # 3. Simulation: 終局までランダムに進めて結果を得る
            score = self._simulate(node.state)

            # 4. Backpropagation: 結果を親ノードに逆伝播させる
            self._backpropagate(node, score)

        # 最も多く探索された子を選ぶ
        choice = max(
            root.children,
            key=lambda child: child.visit_count,
        ).action

        if choice is None:
            raise RuntimeError("No action selected.")

        return self._engine.to_action(choice)

    def _make_node(
        self,
        state: TicTacToeEngineState,
        parent: MCTSNode | None = None,
        action: TicTacToeEngineAction | None = None,
    ) -> MCTSNode:
        node = MCTSNode(
            state=state,
            parent=parent,
            action=action,
        )
        node.untried_actions = list(self._engine.legal_actions(state))
        self._rg.shuffle(node.untried_actions)

        return node

    @staticmethod
    def _uct(
        value_sum: float,
        visits: int,
        parent_visits: int,
        *,
        exploration_weight: float,
    ) -> float:
        if visits == 0:
            return float("inf")

        exploitation = -value_sum / visits
        exploration = exploration_weight * math.sqrt(math.log(parent_visits) / visits)

        return exploitation + exploration

    def _select(self, node: MCTSNode) -> MCTSNode:
        current = node

        while not self._engine.is_terminal(current.state):
            if current.untried_actions:
                break

            current = max(
                current.children,
                key=lambda child: self._uct(
                    child.value_sum,
                    child.visit_count,
                    current.visit_count,
                    exploration_weight=self.exploration_weight,
                ),
            )

        return current

    def _expand(self, node: MCTSNode) -> MCTSNode:
        action = node.untried_actions.pop()

        next_state = self._engine.apply_action(node.state, action)

        child = self._make_node(next_state, node, action)
        node.children.append(child)

        return child

    def _simulate(
        self,
        state: TicTacToeEngineState,
    ) -> float:
        current = state

        # Playout
        while not self._engine.is_terminal(current):
            legal_actions = self._engine.legal_actions(current)
            action = self._rg.choice(legal_actions)
            current = self._engine.apply_action(current, action)

        player = self._engine.current_player(state)
        result = self._engine.result(current)

        if result is not None:
            winner = result.winner

            if winner is player:
                return 1.0
            elif winner is player.opponent:
                return -1.0
            else:
                return 0.0

        raise RuntimeError("Game ended without a result.")

    def _backpropagate(
        self,
        node: MCTSNode,
        score: float,
    ) -> None:
        current = node

        while current is not None:
            current.visit_count += 1
            current.value_sum += score

            score = -score
            current = current.parent

File: src/game_solver/cpp/src/connect_four/mcts.cpp
#include "connect_four/mcts.hpp"

#include <algorithm>
#include <array>
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <limits>
#include <random>
#include <stdexcept>
#include <vector>

namespace connect_four {
namespace {

using NodeIndex = std::size_t;
inline constexpr NodeIndex NO_NODE = std::numeric_limits<NodeIndex>::max();

struct Node {
    EngineState state;
    NodeIndex parent{NO_NODE};
    EngineAction action{};
    std::array<NodeIndex, BOARD_WIDTH> children{};
    std::array<EngineAction, BOARD_WIDTH> untried_actions{};
    std::uint8_t child_count{};
    std::uint8_t untried_count{};
    // Sum of outcomes from state.current_player's perspective.
    double value_sum{};
    std::size_t visit_count{};
};

[[nodiscard]]
Node make_node(const EngineState& state, const NodeIndex parent, const EngineAction action,
               std::mt19937_64& generator) {
    Node node{.state = state, .parent = parent, .action = action};
    if (Engine::is_terminal(state)) {
        return node;
    }
    for (std::uint8_t column = 0; column < BOARD_WIDTH; ++column) {
        if ((state.mask & TOP_MASKS[column]) == 0) {
            node.untried_actions[node.untried_count++] = column;
        }
    }
    std::shuffle(node.untried_actions.begin(), node.untried_actions.begin() + node.untried_count,
                 generator);
    return node;
}

[[nodiscard]]
double uct(const Node& child, const double log_parent_visits,
           const double exploration_weight) noexcept {
    // Selection only sees children already visited by expansion/backpropagation.
    const double visits = static_cast<double>(child.visit_count);
    return -child.value_sum / visits + exploration_weight * std::sqrt(log_parent_visits / visits);
}

[[nodiscard]]
NodeIndex select_node(const std::vector<Node>& nodes, NodeIndex current,
                      const double exploration_weight) noexcept {
    while (!Engine::is_terminal(nodes[current].state) && nodes[current].untried_count == 0) {
        const Node& parent = nodes[current];
        const double log_parent_visits = std::log(static_cast<double>(parent.visit_count));
        NodeIndex best = parent.children[0];
        double best_value = uct(nodes[best], log_parent_visits, exploration_weight);
        for (std::uint8_t i = 1; i < parent.child_count; ++i) {
            const NodeIndex candidate = parent.children[i];
            const double candidate_value =
                uct(nodes[candidate], log_parent_visits, exploration_weight);
            if (candidate_value > best_value) {
                best = candidate;
                best_value = candidate_value;
            }
        }
        current = best;
    }
    return current;
}

[[nodiscard]]
double simulate(EngineState state, std::mt19937_64& generator) noexcept {
    double perspective = 1.0;
    for (;;) {
        if (const auto value = Engine::terminal_value(state)) {
            return perspective * static_cast<double>(*value);
        }

        std::array<EngineAction, BOARD_WIDTH> actions{};
        std::uint8_t count = 0;
        for (std::uint8_t column = 0; column < BOARD_WIDTH; ++column) {
            if ((state.mask & TOP_MASKS[column]) == 0) {
                actions[count++] = column;
            }
        }
        std::uniform_int_distribution<unsigned int> distribution(0, count - 1);
        state = Engine::apply_action_unchecked(state, actions[distribution(generator)]);
        perspective = -perspective;
    }
}

void backpropagate(std::vector<Node>& nodes, NodeIndex current, double score) noexcept {
    while (current != NO_NODE) {
        Node& node = nodes[current];
        ++node.visit_count;
        node.value_sum += score;
        score = -score;
        current = node.parent;
    }
}

}  // namespace

EngineAction mcts(const EngineState& state, const std::size_t simulations,
                  const double exploration_weight, const std::uint64_t seed) {
    if (simulations == 0) {
        throw std::invalid_argument{"connect_four: simulations must be positive"};
    }
    if (!(exploration_weight >= 0.0)) {
        throw std::invalid_argument{"connect_four: exploration_weight must be non-negative"};
    }
    if (Engine::legal_actions(state).empty()) {
        throw std::invalid_argument{"No legal actions available."};
    }

    std::mt19937_64 generator(seed);
    std::vector<Node> nodes;
    nodes.reserve(simulations + 1);
    nodes.push_back(make_node(state, NO_NODE, 0, generator));

    for (std::size_t iteration = 0; iteration < simulations; ++iteration) {
        NodeIndex current = select_node(nodes, 0, exploration_weight);
        if (nodes[current].untried_count != 0) {
            Node& parent = nodes[current];
            const EngineAction action = parent.untried_actions[--parent.untried_count];
            const EngineState next_state = Engine::apply_action_unchecked(parent.state, action);
            const NodeIndex child = nodes.size();
            nodes.push_back(make_node(next_state, current, action, generator));
            nodes[current].children[nodes[current].child_count++] = child;
            current = child;
        }
        backpropagate(nodes, current, simulate(nodes[current].state, generator));
    }

    const Node& root = nodes[0];
    NodeIndex best = root.children[0];
    for (std::uint8_t i = 1; i < root.child_count; ++i) {
        const NodeIndex candidate = root.children[i];
        if (nodes[candidate].visit_count > nodes[best].visit_count) {
            best = candidate;
        }
    }
    return nodes[best].action;
}

}  // namespace connect_four
